File: services/sadi/ckan/lift-ckan.py
#3> <> prov:specializationOf <https://github.com/timrdf/DataFAQs/raw/master/services/sadi/ckan/lift-ckan.py>;
#3>    rdfs:seeAlso <https://github.com/timrdf/DataFAQs/wiki/FAqT-Service>;
#3>    prov:alternateOf <https://github.com/lod-cloud/datahub2void>;
#3> .

import faqt

import sadi
from rdflib import *
import surf

from surf import *
from surf.query import a, select

import rdflib
rdflib.plugin.register('sparql', rdflib.query.Processor,
                       'rdfextras.sparql.processor', 'Processor')
rdflib.plugin.register('sparql', rdflib.query.Result,
                       'rdfextras.sparql.query', 'SPARQLQueryResult')
import httplib
from urlparse import urlparse, urlunparse
import urllib
import urllib2

import ckanclient
import os
import hashlib
import re

import time

# These are the namespaces we are using beyond those already available
# (see http://packages.python.org/SuRF/modules/namespace.html#registered-general-purpose-namespaces)
ns.register(moat='http://moat-project.org/ns#')
ns.register(ov='http://open.vocab.org/terms/')
ns.register(void='http://rdfs.org/ns/void#')
ns.register(con='http://www.w3.org/2000/10/swap/pim/contact#')
ns.register(dcat='http://www.w3.org/ns/dcat#')
ns.register(sd='http://www.w3.org/ns/sparql-service-description#')
ns.register(conversion='http://purl.org/twc/vocab/conversion/')
ns.register(datafaqs='http://purl.org/twc/vocab/datafaqs#')
ns.register(prov='http://www.w3.org/ns/prov#')
ns.register(tag='http://www.holygoat.co.uk/owl/redwood/0.1/tags/')
ns.register(prov='http://www.w3.org/ns/prov#')
ns.register(dcterms='http://purl.org/dc/terms/')
ns.register(foaf='http://xmlns.com/foaf/0.1/')

# The Service itself
class LiftCKAN(faqt.CKANReader):

   # Service metadata.
   label                  = 'lift-ckan'
   serviceDescriptionText = 'Accesses the current CKAN listing for the given datasets and returns a well-designed* RDF form of the same.'
   comment                = 'Thanks to Will Wates\' https://bitbucket.org/okfn/gockan for some modeling guidance.'
   serviceNameText        = 'lift-ckan' # Convention: Match 'name' below.
   name                   = 'lift-ckan' # This value determines the service URI relative to http://localhost:9090/
                                        # Convention: Use the name of this file for this value.
   dev_port = 9225

   def __init__(self):
      # DATAFAQS_PROVENANCE_CODE_RAW_BASE                   +  servicePath  +  '/'  + self.serviceNameText
      # DATAFAQS_PROVENANCE_CODE_PAGE_BASE                  +  servicePath  +  '/'  + self.serviceNameText
      #
      # ^^ The source code location
      #    aligns with the deployment location \/
      #
      #                 DATAFAQS_BASE_URI  +  '/datafaqs/'  +  servicePath  +  '/'  + self.serviceNameText
      faqt.CKANReader.__init__(self, servicePath = 'services/sadi/ckan') # TEMPLATE: change to something like 'services/sadi/faqt/connected/' to get free provenance.

      # Instantiate the CKAN client.
      # http://docs.python.org/library/configparser.html (could use this technique)
      #key = os.environ['X_CKAN_API_Key'] # See https://github.com/timrdf/DataFAQs/wiki/Missing-CKAN-API-Key
      #self.ckan = ckanclient.CkanClient(api_key=key)
      #self.ckan = ckanclient.CkanClient()

   def getOrganization(self):
      result                      = self.Organization()
      result.mygrid_authoritative = True
      result.protegedc_creator    = 'lebot@rpi.edu'
      result.save()
      return result

   def getInputClass(self):
      return ns.DATAFAQS['CKANDataset']

   def getOutputClass(self):
      return ns.DATAFAQS['CKANDataset']

   def process(self, input, output):

      #time.sleep(10)
      print 'processing ' + input.subject

      ckan    = self.getCKANAPI(input)
      ckan_id = self.getCKANIdentiifer(input)
      print 'ckan_id ' + ckan_id

      #
      # GET the current dataset metadata listing from CKAN.
      dataset = {}
      try:
         ckan.package_entity_get(ckan_id)
         dataset = ckan.last_message

         print dataset.keys()
         print
         print dataset

         #
         # Process Attributes
         #
         if 'title' in dataset:
            output.dcterms_title = dataset['title'] # (java'd)

         if 'notes' in dataset:
            output.dcterms_description = dataset['notes'] # (java'd)

         #
         # Process Resources
         #
         Thing   = output.session.get_class(ns.OWL['Thing'])
         Service = output.session.get_class(ns.SD['Service'])
         sparqlEndpoint = None
         for resource in dataset['resources']:
            if resource['format'] == u'api/sparql':
               #
               # <dataset> void:sparqlEndpoint <endpoint> .
               #
               endpoint = Thing(resource['url'])
               endpoint.dcterms_title = resource['description']
               endpoint.save()
               service  = Service('#service-'+hashlib.sha224(resource['url']).hexdigest())
               service.sd_endpoint = endpoint
               service.save()
               sparqlEndpoint = {'url': Thing(resource['url']), # Saved for processing sparql_named_graph below.
                                 'id' : resource['id']} 
               output.void_sparqlEndpoint.append(Service(resource['url']))
            if resource['format'] == u'example/turtle':
               #
               # <dataset> void:exampleResource <> .
               #
               output.void_exampleResource.append(Thing(resource['url']))
               # TODO: elaborate this description.
            else:
               print 'TODO: handle resource: ' + resource['format']

         #
         #
         #
         if 'revision_id' in dataset:
            revisionR = Thing('http://datahub.io/revision/'+dataset['revision_id'])
            revisionR.foaf_isPrimaryTopicOf = output
            if 'metadata_created' in dataset:
               revisionR.dcterms_created = dataset['metadata_created']
            if 'metadata_modified' in dataset:
               revisionR.dcterms_modified = dataset['metadata_modified']
            revisionR.save()

         #
         # Process Groups
         #
         Group = output.session.get_class(ns.DATAFAQS['CKANGroup'])
         for group in dataset['groups']:
            groupR = Group(re.sub('/dataset/.*$','/group/'+group,str(input.subject)))
            print 'group: ' + group + ' -> ' + groupR.subject
            groupR.rdf_type.append(ns.DATAFAQS['CKANGroup'])
            groupR.dcterms_identifier.append(group)
            groupR.rdfs_label.append(group)
            groupR.foaf_name.append(group)
            groupR.save()
            output.dcterms_isPartOf.append(groupR) # (java'd)

         #
         # Process Tags
         #
         Tag = output.session.get_class(ns.TAG['Tag'])
         for tag in dataset['tags']:
            print 'tag: ' + tag
            tagR = Tag('http://datahub.io/tag/' + re.sub(' ','-',tag)) 
            tagR.rdf_type.append(ns.TAG['Tag'])
            tagR.rdfs_label = tag
            tagR.tag_name   = tag
            tagR.save()
            output.tag_taggedWithTag.append(tagR) # (java'd)

         #
         # Process Extras
         #
         DCATDataset  = output.session.get_class(ns.DCAT['Dataset'])
         Distribution = output.session.get_class(ns.DCAT['Distribution'])
         Linkset      = output.session.get_class(ns.VOID['Linkset'])
         links_regex  = re.compile("^links:(.*)$")
         NamedGraph   = output.session.get_class(ns.SD['NamedGraph'])
         CKANDataset  = output.session.get_class(ns.DATAFAQS['CKANDataset'])
         for extra in dataset['extras']:
            if extra == u'triples':
               try:
                  print str(dataset['extras'][extra])
                  triples = long(re.sub('\D','',str(dataset['extras'][extra])))
                  print '     --> ' + str(triples)
                  #
                  # <dataset> void:triples 1000 .
                  #
                  output.void_triples = triples
               except ValueError:
                  # e.g. invalid literal for int() with base 10: '' from rdf-book-mashup
                  print 
                  print 'error on triples'
            elif extra == u'shortname':
               #
               # <dataset> ov:shortName "DBPedia" .
               #
               output.ov_shortName = dataset['extras'][extra] # (java'd)
            elif links_regex.match(str(extra)):
               #
               # <dataset> void:subset [ a void:Linkset; void:target <dataset>, <target>; void:size 50 ] .
               #
               for target in links_regex.findall(str(extra)):
                  target = target.replace(' ','-')
                  targetR = CKANDataset('http://datahub.io/dataset/'+target)
                  targetR.rdf_type.append(ns.DATAFAQS['CKANDataset'])
                  targetR.save()
                  print 'found link to ' + target
                  print '           from ' + output.subject
                  print '                revision_id ' + dataset['revision_id']
                  linkset = Linkset('#linkset-'+target+'-'+hashlib.sha224(output.subject+target+dataset['revision_id']).hexdigest())
                  linkset.void_target.append(output)
                  #linkset.void_target.append(CKANDataset('http://datahub.io/dataset/'+target))
                  linkset.void_target.append(targetR)
                  try:
                     linkset.void_triples = int(dataset['extras'][extra])
                  except ValueError:
                     # e.g. invalid literal for int() with base 10: '???' from scotland-statistical-geography
                     print
                  linkset.save()
                  output.void_subset.append(linkset)
            elif extra == u'preferred_uri':
               #
               # <dataset> con:preferrredURI <another> .
               #
               output.con_preferredURI.append(DCATDataset(dataset['extras'][extra])) # (java'd)
            elif extra == u'namespace':
               #
               # <dataset> void:uriSpace "http://dbpedia.org/resource/" .
               #
               output.void_uriSpace = dataset['extras'][extra] # (java'd)
            elif extra == u'sparql_graph_name':
               #
               # <dataset> dcat:distribution [
               #    a sd:NamedGraph;
               #    sd:name  <http://purl.org/twc/arrayexpress/E-MTAB-104>;
               #    sd:graph <http://datahub.io/en/dataset/arrayexpress-e-afmx-1>;
               #    prov:atLocation <sparqlEndpoint>
               # ];
               #
               if sparqlEndpoint is not None:
                  ng = NamedGraph('#named-graph-'+sparqlEndpoint['id'])
                  ng.sd_name         = Thing(dataset['extras'][extra])
                  ng.prov_atLocation = sparqlEndpoint['url']
                  ng.save()
                  output.dcat_distribution.append(ng) # (java'd)
            else:
               print extra + ' = ' 
               print dataset['extras'][extra]
               print dataset['revision_id']
               print
               print 

         output.rdf_type.append(ns.DCAT['Dataset'])
         output.save()
      except ckanclient.CkanApiNotFoundError:
         print 'Error: could not get ckan dataset ' + ckan_id

# Used when Twistd invokes this service b/c it is sitting in a deployed directory.
resource = LiftCKAN()

# Used when this service is manually invoked from the command line (for testing).
if __name__ == '__main__':
   print resource.name + ' running on port ' + str(resource.dev_port) + '. Invoke it with:'
   print 'curl -H "Content-Type: text/turtle" -d @my.ttl http://localhost:' + str(resource.dev_port) + '/' + resource.name
   sadi.publishTwistedService(resource, port=resource.dev_port)

File: src/java/sadi-services/src/main/java/edu/rpi/tw/data/quality/sadi/ckan/LiftCKAN.java
package edu.rpi.tw.data.quality.sadi.ckan;

import java.util.HashMap;

import org.apache.log4j.Logger;
import org.ckan.CKANException;
import org.ckan.Client;
import org.ckan.Connection;
import org.ckan.Dataset;
import org.ckan.DatasetRevision;
import org.ckan.Extra;
import org.ckan.Group;

import ca.wilkinsonlab.sadi.service.annotations.ContactEmail;
import ca.wilkinsonlab.sadi.service.annotations.InputClass;
import ca.wilkinsonlab.sadi.service.annotations.Name;
import ca.wilkinsonlab.sadi.service.annotations.OutputClass;
import ca.wilkinsonlab.sadi.service.simple.SimpleSynchronousServiceServlet;

import com.hp.hpl.jena.rdf.model.Model;
import com.hp.hpl.jena.rdf.model.Resource;
import com.hp.hpl.jena.sparql.vocabulary.FOAF;
import com.hp.hpl.jena.vocabulary.DC;
import com.hp.hpl.jena.vocabulary.DCTerms;
import com.hp.hpl.jena.vocabulary.RDFS;

import edu.rpi.tw.data.ckan.CKANReader;
import edu.rpi.tw.data.naming.NameFactory;
import edu.rpi.tw.data.rdf.jena.vocabulary.CON;
import edu.rpi.tw.data.rdf.jena.vocabulary.DCAT;
import edu.rpi.tw.data.rdf.jena.vocabulary.DataFAQs;
import edu.rpi.tw.data.rdf.jena.vocabulary.Formats;
import edu.rpi.tw.data.rdf.jena.vocabulary.OV;
import edu.rpi.tw.data.rdf.jena.vocabulary.PROV;
import edu.rpi.tw.data.rdf.jena.vocabulary.Prefixes;
import edu.rpi.tw.data.rdf.jena.vocabulary.SD;
import edu.rpi.tw.data.rdf.jena.vocabulary.SIOC;
import edu.rpi.tw.data.rdf.jena.vocabulary.Tag;
import edu.rpi.tw.data.rdf.jena.vocabulary.VoID;

/**
 * Re-implementation of https://github.com/timrdf/DataFAQs/blob/master/services/sadi/ckan/lift-ckan.py
 * 
 * See also https://github.com/timrdf/DataFAQs/wiki/FAqT-Service
 */
@Name("lift-ckan")
@ContactEmail("lebot@rpi.edu")
@InputClass("http://purl.org/twc/vocab/datafaqs#CKANDataset") 
@OutputClass("http://purl.org/twc/vocab/datafaqs#CKANDataset")
public class LiftCKAN extends SimpleSynchronousServiceServlet {
	
	private static final Logger log = Logger.getLogger(LiftCKAN.class);
	private static final long serialVersionUID = 1L;
	
	private HashMap<String,Resource> ephemeralFormats = new HashMap<String,Resource>();

	/**
	 * 
	 */
	@Override
	public void processInput(Resource input, Resource output) {
        
        try {
            Client client = new Client( new Connection( CKANReader.getCKANAPI(input) ), "");
            //                                          ^ e.g. "http://datahub.io"     <key>
            
        	String inputS = input.asResource().getURI().toString();
        	log.warn("processing "+inputS);
        	
        	String base = CKANReader.getBaseURI(input); // inputS.replaceFirst("\\/dataset\\/.*$", "");
        	
        	log.warn("processing id "+CKANReader.getCKANIdentiifer(input));
        	Dataset ds = client.getDataset(CKANReader.getCKANIdentiifer(input));
        	//                             ^ e.g. "farmers-markets-geographic-data-united-states"
        	
        	Model m = output.getModel();
        	Prefixes.setNsPrefixes(m);

        	// Title
        	tryTitle(ds, output);
        	
        	// Notes
        	if( valid(ds.getNotes()) ) {
        		//output.addProperty(DC.description, ds.getNotes());
        	}
        	
        	//
        	// Resources
        	//
        	Resource sparqlEndpoint = null; // Save for Extras processing named graph.
        	for (org.ckan.Resource resource : ds.getResources() ) {
        		if( valid(resource.getUrl()) ) {
        			
        			Resource distribution = m.createResource(resource.getUrl(), DCAT.Distribution);
        			if( valid(resource.getDescription()) ) {
        				distribution.addProperty(DCTerms.description, resource.getDescription());
        			}
        			output.addProperty(DCAT.distribution, distribution);
        			Resource resourceR = m.createResource(inputS+"/resource/"+resource.getId(), DataFAQs.CKANResource);
        			resourceR.addProperty(DCTerms.isPartOf, output);
        			
					if( "api/sparql".equals(resource.getFormat()) ) {
						
	    				//System.out.println( " name. "  + resource.getName() );
	        			//System.out.println( "    Format: "      + resource.getFormat() );
	        			//System.out.println( "    Mimetype: "    + resource.getMimetype() );
	        			//System.out.println( "    Description: " + resource.getDescription() );
	        			//System.out.println( "    URL: "         + resource.getUrl() + "\n");   
						
						sparqlEndpoint = m.createResource(resource.getUrl(), SD.Service);
	    				output.addProperty(VoID.sparqlEndpoint, sparqlEndpoint);
	    				output.addProperty(DCAT.distribution,   sparqlEndpoint);

	        		}else if( resource.getFormat().startsWith("example/") ) {
	        			/*
	        			 * William Waite's GoLD translation (https://bitbucket.org/okfn/gockan):
	        			 * 
	        			 * <http://thedatahub.org/dataset/farmers-markets-geographic-data-united-states>
	        			 *    dcat:distribution [
	        			 *       a dcat:Distribution ;
						 *       dc:description "Turtle example link" ;
						 *       dc:format [
						 *          moat:taggedWithTag [
						 *              moat:name "example/turtle" ;
						 *              a moat:Tag
						 *          ] ;
						 *          a dc:IMT
						 *      ] ;
						 *      dcat:accessURL <http://logd.tw.rpi.edu/...data-gov-4383-2011-Nov-29.e1.sample.ttl>
						 *    ];
						 * .
	        			 */
	        			distribution.addProperty(DCAT.downloadURL, distribution);

	        			// Format
	        			String format = resource.getFormat().substring("example/".length()).trim();
	        			Resource formatR = Formats.getFormat(format);
	        			if( formatR == null ) {
	        				formatR = m.createResource(DCTerms.FileFormat);
	        			}
        				formatR.addProperty(RDFS.label, format);
        				distribution.addProperty(DCTerms.format, formatR);

	        			output.addProperty(VoID.exampleResource, distribution);
	        		}else if( "csv".equalsIgnoreCase(resource.getFormat()) ) {
	        			distribution.addProperty(DCAT.downloadURL, distribution);
	        			
	        			distribution.addProperty(FOAF.isPrimaryTopicOf, resourceR);
	        			
	        			Resource formatPW  = m.createResource("http://provenanceweb.org/format/mime/text/csv");
	        			formatPW.addLiteral(RDFS.label, "CSV");
	        			formatPW.addLiteral(DCTerms.description, "Comma Separated Values");
	        			
	        			Resource formatPRN = m.createResource("http://provenanceweb.org/formats/pronom/x-fmt/18");
	        			formatPRN.addLiteral(RDFS.label, "CSV");
	        			formatPRN.addLiteral(DCTerms.description, "Comma Separated Values");
	        			
	        			distribution.addProperty(DCTerms.format, formatPW);
	        			distribution.addProperty(DCTerms.format, formatPRN);
	        		}else if( "pdf".equalsIgnoreCase(resource.getFormat()) ) {
	        			distribution.addProperty(DCAT.downloadURL, distribution);
	        			
	        			distribution.addProperty(FOAF.isPrimaryTopicOf, resourceR);
	        			
	        			Resource formatPW  = m.createResource("http://provenanceweb.org/format/mime/application/pdf");
	        			formatPW.addLiteral(RDFS.label, "PDF");
	        			formatPW.addLiteral(DCTerms.description, "Portable Document Format");
	        			
	        			Resource formatPRN = m.createResource("http://provenanceweb.org/formats/pronom/fmt/276");
	        			formatPRN.addLiteral(RDFS.label, "PDF");
	        			formatPRN.addLiteral(DCTerms.description, "Portable Document Format");
	        			
	        			distribution.addProperty(DCTerms.format, formatPW);
	        			distribution.addProperty(DCTerms.format, formatPRN);
	        		}else if( "pptx".equalsIgnoreCase(resource.getFormat()) ) {
	        			distribution.addProperty(DCAT.downloadURL, distribution);
	        			
	        			distribution.addProperty(FOAF.isPrimaryTopicOf, resourceR);
	        			
	        			Resource formatPW  = m.createResource("http://provenanceweb.org/format/mime/application/vnd.ms-powerpoint");
	        			formatPW.addLiteral(RDFS.label, "PDF");
	        			formatPW.addLiteral(DCTerms.description, "Microsoft Powerpoint");
	        			
	        			Resource formatPRN = m.createResource("http://provenanceweb.org/formats/pronom/fmt/215");
	        			formatPRN.addLiteral(RDFS.label, "PDF");
	        			formatPRN.addLiteral(DCTerms.description, "Microsoft Powerpoint for Windows");
	        			
	        			distribution.addProperty(DCTerms.format, formatPW);
	        			distribution.addProperty(DCTerms.format, formatPRN);
	        		}else {
	        			distribution.addProperty(FOAF.isPrimaryTopicOf, resourceR);
	        			distribution.addProperty(DCAT.downloadURL, distribution);
	        			
	        			Resource format = getFormat(resource.getFormat(), m);
	        			
	        			distribution.addProperty(DCTerms.format, format);
	        		}
        		}
        	}

        	
        	//
        	// Groups
        	//
        	for( Group group : ds.getGroups() ) {
    			Resource groupR = m.createResource(base+"/group/"+group.getName().replaceAll(" ", "-"), DataFAQs.CKANGroup);
    			groupR.addLiteral(DC.identifier, group.getName().replaceAll(" ", ""));
    			groupR.addLiteral(FOAF.name, group.getName());
    			groupR.addLiteral(RDFS.label, group.getName());
    			output.addProperty(DCTerms.isPartOf, groupR);
        	}
        	
        	
        	//
        	// Tags
        	//
        	for( org.ckan.Tag tag : ds.getTags() ) {
        		//output.addProperty(edu.rpi.tw.data.rdf.jena.vocabulary.Tag.taggedWithTag,
        		//		tag.getDisplayName() + " - " + tag.getId() + " - " + tag.getName() + " - " + tag.getVocabularyId());
        		
    			Resource tagR = m.createResource(base+"/tag/"+tag.getName().replaceAll(" ", "-"), Tag.Tag);
    			output.addProperty(Tag.taggedWithTag, tagR);
    			tagR.addLiteral(RDFS.label, tag.getDisplayName()); // Tag.name
    			
    			if( tag.getName().startsWith("format-") ) {
    				// TODO assert VoID.vocabulary based on prefix.cc
    			}
        	}
        	
        	
        	//
        	// Extras
        	//
        	for( Extra extra : ds.getExtras() ){
        		if( "triples".equals(extra.getKey()) ) {
        			long size = Long.parseLong(extra.getValue().replaceAll("[^0-9]", ""));
        			output.addLiteral(VoID.triples, size);
        		}else if( "shortname".equals(extra.getKey()) && valid(extra.getValue()) ) {
        			output.addLiteral(OV.shortName, trim(extra.getValue()));
        		}else if( extra.getKey().startsWith("links:") ) {
        			String otherID  = extra.getKey().substring("links:".length()).replaceAll(" ", "-");
        			String other    = inputS.replaceFirst("\\/dataset\\/.*$", "/dataset/"+otherID);    
        			Resource otherR = m.createResource(other, DataFAQs.CKANDataset);
        			
        			// Jena can't handle relative fragIDs...
        			/*Resource linkset = m.createResource("#linkset-"+otherID+"-"+
        												NameFactory.getMD5(inputS+otherID+ds.getRevision_id()),
        												VoID.Linkset);*/
        			Resource linkset = m.createResource(NameFactory.INSTANCE_HUB+"linkset/"+otherID+"/"+
														NameFactory.getMD5(inputS+otherID+ds.getRevision_id()),
														VoID.Linkset);
        			long size = Long.parseLong(extra.getValue().replaceAll("[^0-9]", ""));
        			linkset.addLiteral(VoID.triples, size);
        			linkset.addProperty(VoID.target, otherR);
        			linkset.addProperty(VoID.target, output);
        			
        			log.warn("void:subset "+linkset.getURI().toString());
        			output.addProperty(VoID.subset, linkset);
        		}else if( "preferred_uri".equals(extra.getKey()) && valid(extra.getValue()) ) {
        			output.addProperty(CON.preferredURI, m.createResource(trim(extra.getValue())));
        			//output.addProperty(DataFAQs.todo, "extra: " + extra.getKey() + " = " + extra.getValue());
        		}else if( "namespace".equals(extra.getKey()) && valid(extra.getValue()) ) {
        			output.addLiteral(VoID.uriSpace, trim(extra.getValue()));
        			//output.addProperty(DataFAQs.todo, "extra: " + extra.getKey() + " = " + extra.getValue());
        		}else if( "sparql_graph_name".equals(extra.getKey()) && valid(extra.getValue()) ) {
        			String name = trim(extra.getValue());
                    /*
                    *   <dataset> dcat:distribution [
                    *      a sd:NamedGraph;
                    *      sd:name  <http://purl.org/twc/arrayexpress/E-MTAB-104>;
                    *      sd:graph <http://datahub.io/en/dataset/arrayexpress-e-afmx-1>;
                    *      prov:atLocation <sparqlEndpoint>
                    *   ];
                    */
        			// Jena can't handle relative fragIDs...
        			/*Resource namedGraph = sparqlEndpoint != null ? 
        						m.createResource( "#named-graph-"+NameFactory.getMD5(sparqlEndpoint.getURI().toString()+name), SD.NamedGraph):
            					m.createResource( "#named-graph-"+NameFactory.getMD5(                                   name), SD.NamedGraph);*/
        			Resource namedGraph = sparqlEndpoint != null ? 
    						m.createResource(NameFactory.INSTANCE_HUB+"named-graph/"+NameFactory.getMD5(sparqlEndpoint.getURI().toString()+name), SD.NamedGraph):
        					m.createResource(NameFactory.INSTANCE_HUB+"named-graph/"+NameFactory.getMD5(                                   name), SD.NamedGraph);   
        			output.addProperty(DCAT.distribution, namedGraph);
        			namedGraph.addProperty(SD.name, m.createResource(name));
        			if( sparqlEndpoint != null ) {
        				namedGraph.addProperty(PROV.atLocation, sparqlEndpoint);
        			}
        		}else {
        			output.addProperty(DataFAQs.todo, "extra: " + extra.getKey() + " = " + extra.getValue());
        		}
        	}
        
        	//
        	// Dataset entry revisions
        	//
        	for( DatasetRevision revision : client.getDatasetRevisions(CKANReader.getCKANIdentiifer(input)) ) {
        		Resource revisionR = m.createResource(base+"/revision/"+revision.getID());
        		
        		Resource user = null;
        		if( revision.getAuthor() != null && revision.getAuthor().startsWith("http") ) {
        			user = m.createResource(revision.getAuthor());
        		}else if(revision.getAuthor() != null && 
        				revision.getAuthor().matches(
        				"^(([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])\\.){3}([0-9]|[1-9][0-9]|1[0-9]{2}|2[0-4][0-9]|25[0-5])$") ) {
        			user = m.createResource(base+"/user/"+revision.getAuthor());
        			revisionR.addLiteral(SIOC.ip_address, revision.getAuthor());
        		}else if( revision.getAuthor() != null ){
        			user = m.createResource(base+"/user/"+revision.getAuthor());
        		}
        		
        		output.addProperty(PROV.wasDerivedFrom, revisionR);
        		revisionR.addProperty(DCTerms.created, revision.getTimestamp());
        		if( revision.getMessage() != null && revision.getMessage().length() > 0 ) {
        			revisionR.addProperty(DCTerms.description, revision.getMessage());
        		}
        		if( user != null ) {
        			revisionR.addProperty(PROV.wasAttributedTo, user);
            		user.addProperty(FOAF.name, revision.getAuthor());
        		}
        	}
        	//output.addProperty(RDF.type, DCAT.Dataset);
        }catch ( CKANException e ) {
            log.error("CKANException " + e.getMessage());
        }
	}
	
	private Resource getFormat(String label, Model m) {
		Resource formatR = Formats.getFormat(label);
		if( formatR == null ) {
			if( !ephemeralFormats.containsKey(label) ) {
				formatR = m.createResource(DCTerms.FileFormat);
				ephemeralFormats.put(label, formatR);
			}
			formatR = ephemeralFormats.get(label);
		}
		formatR.addProperty(RDFS.label, label);
		formatR.addProperty(DataFAQs.todo, "URI for FileFormat " + label);
		return formatR;
	}
	
	private void tryTitle(Dataset ds, Resource output) {
		try {
	    	// Title
	    	if( valid(ds.getTitle()) ) {
	    		output.addProperty(DC.title, ds.getTitle());
	    	}
		} catch(Exception e) {
			log.error("exception on title");
		}
	}
	
	private boolean valid(String string) {
		return string != null && string.length() > 0;
	}
	
	private String trim(String string) {
		return valid(string) ? string.replaceAll("^\"", "").replaceAll("\"$", "") : string;
	}
}